Modules/_ctypes/cfield.c

```diff
@@ -61,238 +72,146 @@ Py_ssize_t LOW_BIT(Py_ssize_t offset);
 
-MSVC mode works similar except for bitfield packing.  Adjacent bit-fields are
-packed into the same 1-, 2-, or 4-byte allocation unit if the integral types
-are the same size and if the next bit-field fits into the current allocation
-unit without crossing the boundary imposed by the common alignment requirements
-of the bit-fields.
+/*[clinic input]
+@classmethod
+_ctypes.CField.__new__ as PyCField_new
 
-See https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#index-mms-bitfields for details.
+    name: object(subclass_of='&PyUnicode_Type')
+    type as proto: object
+    size: Py_ssize_t
+    offset: Py_ssize_t
+    index: Py_ssize_t
+    bit_size as bit_size_obj: object = None
 
-We do not support zero length bitfields.  In fact we use bitsize != 0 elsewhere
-to indicate a bitfield. Here, non-bitfields need bitsize set to size*8.
+[clinic start generated code]*/
 
-PyCField_FromDesc manages:
-- *psize: the size of the structure / union so far.
-- *poffset, *pbitofs: 8* (*poffset) + *pbitofs points to where the next field
-  would start.
-- *palign: the alignment requirements of the last field we placed.
-*/
-
-static int
-PyCField_FromDesc_gcc(Py_ssize_t bitsize, Py_ssize_t *pbitofs,
-                Py_ssize_t *psize, Py_ssize_t *poffset, Py_ssize_t *palign,
-                CFieldObject* self, StgInfo* info,
-                int is_bitfield
-                )
-{
-    // We don't use poffset here, so clear it, if it has been set.
-    *pbitofs += *poffset * 8;
-    *poffset = 0;
-
-    *palign = info->align;
-
-    if (bitsize > 0) {
-        // Determine whether the bit field, if placed at the next free bit,
-        // fits within a single object of its specified type.
-        // That is: determine a "slot", sized & aligned for the specified type,
-        // which contains the bitfield's beginning:
-        Py_ssize_t slot_start_bit = round_down(*pbitofs, 8 * info->align);
-        Py_ssize_t slot_end_bit = slot_start_bit + 8 * info->size;
-        // And see if it also contains the bitfield's last bit:
-        Py_ssize_t field_end_bit = *pbitofs + bitsize;
-        if (field_end_bit > slot_end_bit) {
-            // It doesn't: add padding (bump up to the next alignment boundary)
-            *pbitofs = round_up(*pbitofs, 8*info->align);
-        }
-    }
-    assert(*poffset == 0);
-
-    self->offset = round_down(*pbitofs, 8*info->align) / 8;
-    if(is_bitfield) {
-        Py_ssize_t effective_bitsof = *pbitofs - 8 * self->offset;
-        self->size = BUILD_SIZE(bitsize, effective_bitsof);
-        assert(effective_bitsof <= info->size * 8);
-    } else {
-        self->size = info->size;
+static PyObject *
+PyCField_new_impl(PyTypeObject *type, PyObject *name, PyObject *proto,
+                  Py_ssize_t size, Py_ssize_t offset, Py_ssize_t index,
+                  PyObject *bit_size_obj)
+/*[clinic end generated code: output=43649ef9157c5f58 input=3d813f56373c4caa]*/
+{
+    CFieldObject* self = NULL;
+    if (size < 0) {
```

Comment:
Does it make sense to have `size = 0`?

Reply:
I'd... prefer to not answer :)
The important thing is that `ctypes` allows it.

Reply:
Ok, this answer satisfies me. I was just worried that `size = 0` wasn't allowed at all (intuitively I'd say that's it's a non-field, so it would be equivalent to not specifying it but let's not dive into this issue now).